EAMxx: Adds aerosols heterogeneous freezing calculations in P3 microphysics

singhbalwinder opened the pull request:
The heterogeneous freezing calculations from prognostic aerosols are 
added to P3 microphysics. Setting `use_hetfrz_classnuc` to `true` 
will turn on these calculations. Otherwise, P3 will use the default 
prescribed aerosol calculations.

Feature description by Kai Zhang (@kaizhangpnl ):

The ice nucleation parameterization in E3SM is based on Classical 
Nucleation Theory (CNT) and accounts for the impact of aerosols 
on heterogeneous ice formation in mixed-phase clouds. It describes 
the activation of ice-nucleating particles (INPs) under supersaturated
 conditions, incorporating the energy barrier required for ice embryo 
formation on aerosol surfaces. The CNT scheme in E3SM depends 
on both interstitial and cloud-borne black carbon and dust aerosols
 (Wang et al., 2014).
The parameterization includes three heterogeneous ice nucleation
 mechanisms:
1. Deposition nucleation – Direct deposition of water vapor onto an
 INP, forming ice without an intermediate liquid phase.
2. Immersion freezing – Ice formation within supercooled cloud 
droplets containing INPs.
3. Contact freezing – Freezing triggered by the collision of an INP 
with a supercooled droplet.
This scheme is coupled with the Modal Aerosol Module (MAM4) to 
account for the predicted aerosol properties and the influences cloud 
microphysics by regulating heterogeneous ice nucleation in the mixed-phase 
clouds.

Wang, Y., Liu, X., Hoose, C., and Wang, B.: Different contact angle 
distributions for heterogeneous ice nucleation in the Community 
Atmospheric Model version 5, Atmos. Chem. Phys., 14, 
10411–10430, https://doi.org/10.5194/acp-14-10411-2014, 2014.

[BFB] for EAM and EAMxx

singhbalwinder commented:
TODO:
1. Turn off this feature for default EAMxx
2. Revive commented-out P3 tests after adding missing arguments in various function signatures.

bartgol commented on components/eamxx/src/physics/p3/tests/p3_main_unit_tests.cpp:
```diff
@@ -212,15 +212,15 @@ void run_bfb_p3_main_part2()
   // Get data from cxx
   for (auto& d : isds_cxx) {
-    p3_main_part2_host(
+    /*p3_main_part2_host(
```
Why is this test commented?

bartgol commented on components/eamxx/src/physics/p3/tests/p3_nc_conservation_tests.cpp:
```diff
@@ -57,7 +57,7 @@ struct UnitWrap::UnitTest<D>::TestNcConservation : public UnitWrap::UnitTest<D>:
       }
 
-      Functions::nc_conservation(nc, nc_selfcollect_tend, cxx_device(offset).dt, nc_collect_tend, nc2ni_immers_freeze_tend, nc_accret_tend, nc2nr_autoconv_tend);
+      //Functions::nc_conservation(nc, nc_selfcollect_tend, cxx_device(offset).dt, nc_collect_tend, nc2ni_immers_freeze_tend, nc_accret_tend, nc2nr_autoconv_tend);
```
Why is this commented?

bartgol replied:
Same question for all the other unit tests.

singhbalwinder replied:
Thanks, @mahf708 and Luca for your reviews. I commented out these tests because the function signatures changed when I added this feature, and they failed. @overfelt is already working on reviving these tests (see my comment above about TODOs). He has revived some of the tests already. We will uncomment all these tests and revive them again before changing the status of this PR from "draft" to normal.

bartgol reviewed (commented):
I have a few comments. Mostly: why are lots of unit tests now commented?

bartgol reviewed (commented):
I have a few comments. Mostly: why are lots of unit tests now commented?

singhbalwinder commented on components/eamxx/src/physics/p3/eamxx_p3_process_interface.cpp:
```diff
@@ -44,6 +44,7 @@ void P3Microphysics::set_grids(const std::shared_ptr<const GridsManager> grids_m
   infrastructure.kte = m_num_levs-1;
   infrastructure.predictNc = m_params.get<bool>("do_predict_nc",true);
   infrastructure.prescribedCCN = m_params.get<bool>("do_prescribed_ccn",true);
+  infrastructure.use_hetfrz_classnuc = m_params.get<bool>("use_hetfrz_classnuc",true);
```
We will revisit and mimic how the `do_ice_production` flag is used.

singhbalwinder commented on components/eamxx/src/physics/p3/impl/p3_CNT_couple_impl.hpp:
```diff
@@ -0,0 +1,57 @@
+      ninuc_cnt.set(mask, frzdep*1.0e6/rho, Zero);
+      qinuc_cnt.set(mask, ninuc_cnt*mi0, Zero);
+      break;
+    default:
```
Thanks. We should add `EKAT_KERNEL_ERROR(...)` as in the current logic, its value can only be 1 or 2. If it has some other value, it should error out.

singhbalwinder replied:
I have added the `EKAT_KERNEL_ERROR`, the other comment will be addressed when we do some restructuring of P3 in a future PR.

singhbalwinder commented on components/eamxx/src/physics/p3/impl/p3_CNT_couple_impl.hpp:
```diff
@@ -0,0 +1,57 @@
+  const auto mask = qc_incld > qsmall;
+  switch (Iflag) {
+    case 1:  // cloud droplet immersion freezing
+      ncheti_cnt.set(mask, frzimm*1.0e6/rho /* frzimm input is in [#/cm3] */ , Zero);
```
That is a good question. I am not sure about that. @kaizhangpnl or @AaronDonahue might know if `mask` can often be false or not.

singhbalwinder commented on components/eamxx/cime_config/namelist_defaults_scream.xml:
```diff
@@ -192,6 +192,7 @@ be lost if SCREAM_HACK_XML is not enabled.
     <!-- P3 microphysics -->
     <p3 inherit="atm_proc_base">
       <do_prescribed_ccn>true</do_prescribed_ccn>
+      <use_hetfrz_classnuc>true</use_hetfrz_classnuc>
```
Sure. We will add the following info:
```
doc=“Switch to turn on heterogeneous freezing due to prognostic aerosols”
type=”logical”
```

AaronDonahue commented on components/eamxx/src/physics/p3/impl/p3_CNT_couple_impl.hpp:
```diff
@@ -0,0 +1,57 @@
+{
+  constexpr Scalar pi       = C::Pi;
+  constexpr Scalar rho_h2o  = C::RHO_H2O;
+  constexpr Scalar qsmall   = 1.0e-18; // BAD_CONSTANT!
```
Could you add instead a string "TODO: ..." ?

singhbalwinder replied:
Yes, we can change it to a TODO.

singhbalwinder commented on components/eamxx/src/physics/p3/impl/p3_main_impl_part2.hpp:
```diff
@@ -451,8 +470,8 @@ ::p3_main_part2(
-      rho_qm_cloud, th_atm(k), qv(k), qi(k), ni(k), qm(k), bm(k), qc(k),
-      nc(k), qr(k), nr(k), not_skip_all);
+      rho_qm_cloud, ncheti_cnt, nicnt, ninuc_cnt, qcheti_cnt, qicnt, qinuc_cnt,  th_atm(k), qv(k), qi(k), ni(k), qm(k), bm(k), qc(k),
+      nc(k), qr(k), nr(k), use_hetfrz_classnuc, not_skip_all);
```
Thanks, Naser. If I understood correctly, your method would expose the boolean earlier in the logic, which, I agree, is an advantage. However, if we overload the functions, wouldn't that lead to significant code duplication, given that the boolean only affects a few lines? To avoid code duplication, we may need to re-structure the logic by introducing helper functions for the shared code. 

I initially considered using default function arguments to reduce duplication, but that would also require some restructuring. Perhaps @bartgol has a better approach.

I also want to emphasize that we need this code for urgent simulations. If implementing this change requires extensive restructuring, I suggest addressing it in a follow-up PR. Thanks!

bartgol replied:
**tl;dr** : I would leave as is.


I agree that having to pass lots of views (the signatures of some P3 functions are ridicolously long) just in case you need them seems silly, especially if the call site already knows it's not going to need them. Lots of interfaces need to be adjusted, so that they can propagate the extra args to the calls down the stack.

Alas, that's how P3 works now. A better design would be to NOT pass (a lot of) views, but a handful of self-explanatory structures. We do that for p3 main (with `P3Infrastructure`, `P3DiagnosticOutputs`, etc), but we don't use it in any of the inner functions. Imho, it's defying the whole purpose, since we create these structures just to undo them right inside the p3 main call. There's a reason for it, ofc, which is that these structures have views with the col index, while inner fcns all work on an individual column.

Imho, we should at some point rewrite the p3 interfaces (all of them), since maintaining/debugging them is a bit of a nightmare for ppl who don't know p3 very well already. But that should not happen in this PR, clearly.

singhbalwinder commented on components/eamxx/src/physics/p3/impl/p3_main_impl.hpp:
```diff
@@ -243,6 +248,7 @@ ::p3_main_internal(
 
     p3_main_part2(
       team, nk_pack, runtime_options.max_total_ni, infrastructure.predictNc, infrastructure.prescribedCCN, infrastructure.dt, inv_dt,
+      runtime_options.use_hetfrz_classnuc, ohetfrz_immersion_nucleation_tend, ohetfrz_contact_nucleation_tend, ohetfrz_deposition_nucleation_tend,
```
Good catch! Thanks. I didn't notice the `runtime_options` struct is sent as an arg. I will revisit this to fix it. I will change the variable names as well.

singhbalwinder commented on components/eamxx/src/physics/p3/eamxx_p3_process_interface.cpp:
```diff
@@ -91,6 +91,23 @@ void P3Microphysics::set_grids(const std::shared_ptr<const GridsManager> grids_m
+  // units of number mixing ratios of tracers
+  constexpr auto frz_unit = 1 / (cm * cm * cm * s);
+  //  heterogeneous freezing by immersion nucleation [cm^-3 s^-1]
+  add_field<Required>("hetfrz_immersion_nucleation_tend", scalar3d_layout_mid,
+                      frz_unit, grid_name, ps);
+
+  // heterogeneous freezing by contact nucleation [cm^-3 s^-1]
+  add_field<Required>("hetfrz_contact_nucleation_tend", scalar3d_layout_mid, frz_unit,
+                      grid_name, ps);
+
+  // heterogeneous freezing by deposition nucleation [cm^-3 s^-1]
+  add_field<Required>("hetfrz_deposition_nucleation_tend", scalar3d_layout_mid,
+                      frz_unit, grid_name, ps);
```
When MAM is inactive, values for these variables will be picked up from the namelist and will stay the same for the entire simulation. The simulation will run as usual with no impact from these variables.

I can hide it behind an if/else if it is not the desired behavior.  Is there a way for P3 to know if MAM4 is active or not? Previously, when I discussed this, the design principle did not allow this, as each parameterization should be able to run independently without the knowledge of other parameterizations/processes. Otherwise, it makes the logic complex (e.g., to decide if different combinations of processes are valid or not). If it has changed, please let me know, and I will add if/else blocks.

singhbalwinder commented:
@mahf708 and @bartgol : I have now addressed all the review comments. Please let me know if there is anything still missing. The P3 tests passed on Compy. I will try running the tests on PM-GPUs.

singhbalwinder commented:
Thanks, Naser! With your help, I have fixed this test.

bartgol commented:
> 1. Your PR is making the p3 unit tests fail (these are p3_tests and p3_sk_tests, found under components/eamxx/p3/tests). The change seems to be in the comparison only, so something you did is changing those. That aligns with my prior experience, but I opted to close the PR rather than figuring it out. See link below.

@singhbalwinder  I noticed that only p3_tests/p3_sk_tests fail, while all of the XYZ_baseline_cmp tests (where XYZ includes p3) pass. Is it b/c you hard code the new fields to the constant value they have in master?

bartgol commented:
> Update the MMF2 test fails with this annoying error:
> 
> ```
> e3sm.exe: /home/runner/_work/E3SM/E3SM/externals/ekat/src/ekat/kokkos/ekat_subview_utils.hpp:32: ekat::Unmanaged<Kokkos::View<ST*, Kokkos::LayoutRight, Props ...> > ekat::subview(ViewLR<ST**, Props ...>&, int) [with ST = const Pack<double, 1>; Props = {Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0>}; Unmanaged<Kokkos::View<ST*, Kokkos::LayoutRight, Props ...> > = Kokkos::View<const Pack<double, 1>*, Kokkos::LayoutRight, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1> >; ViewLR<ST**, Props ...> = Kokkos::View<const Pack<double, 1>**, Kokkos::LayoutRight, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >]: Assertion `v.data() != nullptr' failed.
> 
> Program received signal SIGABRT: Process abort signal.
> ```
> 
> This type of error is almost certainly to do with missing views in the diagnostic_inputs struct based on my prior experience, but I could be misremembering things. Could be fixed by populating these in the PAM interface.

It could be a bad index of a subview. E.g., try to subview (ncols,nlevs) at first index ncols ...

singhbalwinder commented:
Both p3_tests/p3_sk_tests passed on my end on Compy (standalone build) and pm-gpu (using `test-all-scream`). Yesterday I tried with different pack sizes and omp threads on Compy and they all passed. I do not expect any non-MAM4xx tests to fail as the new flag is set to `false` by default.

In p3 tests, I am using the `engine` to generate input for the new flag. I thought it was okay to use `engine` as the asserts will compare the outputs consistently (e.g. output when flag is true on device vs. output when flag is true on host and vice versa). Should I hardwire it to false always?

Once I reproduce it locally, I should be able to debug it. I am currently looking at ways to reproduce it.

brhillman commented on components/eamxx/src/physics/p3/impl/p3_conservation_impl.hpp:
```diff
@@ -28,7 +35,13 @@ ::cloud_water_conservation(const Spack& qc, const Scalar dt,
     qc2qi_collect_tend.set(enforce_conservation, qc2qi_collect_tend*ratio);
-    qc2qi_hetero_freeze_tend.set(enforce_conservation, qc2qi_hetero_freeze_tend*ratio);
+    if(use_hetfrz_classnuc){
+         qcheti_cnt.set(enforce_conservation, qcheti_cnt*ratio);
```
Why are these named differently? Are these meant to correspond to the concentration tendencies like the rest of this block?

singhbalwinder replied:
If I understand your question correctly, these variables store quantities to compute heterogeneous ice formation using CNT. I followed the naming convention and logic used in SCREAMv0 P3 and E3SMv3 source codes. Please let me know if I misunderstood your questions.

singhbalwinder commented:
I have added the feature description provided by @kaizhangpnl in the PR message. This PR passes all the EAMxx tests. There is only MMF2 test failing, but that test is failing with the same error message using master too. It is now ready to merge.
